File: js/templates/lead-pipeline/src/dedupe.ts
export interface Lead {
  email?: string
  [key: string]: unknown
}

/** Dedupe leads by normalized email; drops rows without an email. Pure + testable. */
export function dedupeByEmail(leads: Lead[]): Lead[] {
  const seen = new Set<string>()
  const out: Lead[] = []
  for (const lead of leads) {
    const key = (lead.email ?? "").trim().toLowerCase()
    if (!key || seen.has(key)) continue
    seen.add(key)
    out.push(lead)
  }
  return out
}

File: python/templates/lead-pipeline/app/dedupe.py
from typing import Any, Dict, List


def dedupe_by_email(leads: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
    """Dedupe leads by normalized email; drops rows without an email. Pure + testable."""
    seen = set()
    out: List[Dict[str, Any]] = []
    for lead in leads:
        key = (lead.get("email") or "").strip().lower()
        if not key or key in seen:
            continue
        seen.add(key)
        out.append(lead)
    return out
